### PR TITLE
[threads] Have MonoInternalThread to be passed to mono_thread_internal_detach

### PR DESCRIPTION
**Pull request by luhenry**

### CODE REVIEW

File: mono/metadata/threads.c [hunk 1]
```diff
@@ -691,6 +691,8 @@ mono_thread_attach_internal (MonoThread *thread, gboolean force_attach, gboolean
 		return FALSE;
 	}
 
+	mono_thread_info_set_detach_user_data (info, (gpointer) internal);
```
Comment (kumpera) [inline mono/metadata/threads.c]: This is storing a managed pointer into the native heap.

File: mono/utils/mono-threads.c [hunk 2]
```diff
@@ -353,147 +353,6 @@ thread_handle_destroy (gpointer data)
 	g_free (thread_handle);
 }
 
-static void*
-register_thread (MonoThreadInfo *info, gpointer baseptr)
```
Comment (kumpera) [inline mono/utils/mono-threads.c]: If this really needed? It will introduce a lot of history noise for no particular reason.

File: configure.ac [hunk 3]
```diff
@@ -4015,12 +4015,6 @@ if test "x$DISABLE_MCS_DOCS" = "xdefault"; then
    DISABLE_MCS_DOCS=$DISABLE_MCS_DOCS_default
 fi
 
-AC_ARG_WITH(lazy_gc_thread_creation, [  --with-lazy-gc-thread-creation=yes|no      Enable lazy runtime thread creation, embedding host must do it explicitly (defaults to no)],[
```
Comment (kumpera) [inline configure.ac]: Why is this being removed? What's the justification for that?

Comment (luhenry) [inline configure.ac]: It's dead: if you define `LAZY_GC_THREAD_CREATION` then no finalizer thread will ever be created.

File: mono/utils/mono-threads.c (hunk 2 above)
Comment (luhenry) [inline mono/utils/mono-threads.c]: I will revert that

File: mono/metadata/threads.c (hunk 1 above)
Comment (luhenry) [inline mono/metadata/threads.c]: I am not very happy with this solution, so I am exploring others. If we decide to stick with this one, I will have to register a GC root in MonoThreadInfo.

File: configure.ac (hunk 3 above)
Comment (kumpera) [inline configure.ac]: How did you determine it was dead?
AFAICT, we still ship it: https://github.com/xamarin/xamarin-macios/blob/0a47a8ce0c6d00e7d36e3315289c99de01f417b2/builds/Makefile#L1263

Comment (luhenry) [inline configure.ac]: You are right, it's still in use. I am going to add it back to a header, as it's not defined in any of them.

Comment (kumpera) [inline configure.ac]: By moving to a header the idea is to make this conditional to the host being iOS?
iOS functions fine without it as it's just an optimization.
How does it get better without the configure flag?

**Comment by luhenry** [comment]
Replaced by https://github.com/mono/mono/pull/5058

**Pull request closed without merging**